random cleanups

neheb opened the pull request.

neheb committed:
Return directly instead of using a temporary variable

neheb committed:
string_view conversions

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
replace for loop with std::copy

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use to_string

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use back()

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
optimize propertyInfo slightly

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
don't return by addressof for pointers

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
avoid strncmp

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use std::equal instead of a manual loop

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove redundant inherited copy functions

These are no longer needed with modern compilers.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use enum for TiffType

simpler

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
meson: update subprojects

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang warnings

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented on meson.build:
```diff
@@ -17,6 +17,7 @@ else
 endif
 
 if host_machine.system() == 'windows'
+  add_project_arguments('-D_WIN32_WINNT=0x0600', language: 'cpp')
```
Good point. Will remove

neheb committed:
add some missing stringFormat conversions

Signed-off-by: Rosen Penev <rosenp@gmail.com>

kevinbackhouse commented on samples/geotag.cpp:
```diff
@@ -109,7 +110,9 @@ class Position {
 
   Position() = default;
-  virtual ~Position() = default;
+  ~Position() = default;
```
Can you make the whole class `final`? Otherwise, I think a non-virtual destructor is considered bad practice, isn't it?

neheb replied:
Nothing is being inherited here.

kevinbackhouse commented on samples/geotag.cpp:
```diff
@@ -260,7 +263,7 @@ struct UserData final {
 
 // XML Parser Callbacks
-static void startElement(void* userData, const char* name, const char** atts) {
+void startElement(void* userData, const char* name, const char** atts) {
```
Why did you remove the `static` here?

neheb replied:
Redundant in anonymous namespace

kevinbackhouse commented on src/properties.cpp:
```diff
@@ -5033,8 +5033,8 @@ const XmpPropertyInfo* XmpProperties::propertyInfo(const XmpKey& key) {
   if (auto i = property.find_last_of('/'); i != std::string::npos) {
-    for (; i != std::string::npos && !isalpha(property.at(i)); ++i) {
-    }
+    while (!std::isalpha(property.at(i)))
```
This should probably check for `\0`. I agree that the previous `i != std::string::npos` check was wrong.

kevinbackhouse commented on src/tiffcomposite_int.cpp:
```diff
@@ -557,34 +557,30 @@ TiffComponent* TiffComponent::doAddChild(UniquePtr /*tiffComponent*/) {
-  TiffComponent* tc = tiffComponent.release();
-  components_.push_back(tc);
-  return tc;
+  components_.push_back(tiffComponent.release());
```
Some of these changes will have merge conflicts with my changes in #3179.

kevinbackhouse commented on src/value.cpp:
```diff
@@ -133,20 +133,16 @@ DataValue* DataValue::clone_() const {
-    os << static_cast<int>(value_.at(i));
-    if (i < end - 1)
-      os << " ";
+  if (!value_.empty()) {
+    std::copy(value_.begin(), value_.end() - 1, std::ostream_iterator<int>(os, " "));
+    os << static_cast<int>(value_.back());
```
This seems a weird way to write the code. Is it a standard idiom that I'm not familiar with? If so, please could you add a comment with a hyperlink to the relevant wikipedia page or similar?

neheb replied:
I don't think it's standard. The goal way to avoid a manual loop.

neheb committed:
remove std::swap workaround

std::swap was added here to avoid a use after move warning. But in
reality, the second usage of object is wrong and should be changed to
nullptr.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Rebased.
